# 01. Exploratory Data Analysis (EDA) — MIMIC-IV-ED

This notebook contains the initial data exploration of the MIMIC-IV Emergency Department dataset.  
The goal is to understand the structure, volume, and quality of each table before any preprocessing or modeling.

**Tables available:**
- `edstays` — Core table: one row per ED visit
- `triage` — Vital signs and chief complaint recorded at patient arrival
- `diagnosis` — ICD diagnosis codes assigned during the stay
- `medrecon` — Medication reconciliation (prior medications)
- `pyxis` — Medications dispensed inside the ED
- `vitalsign` — Repeated vital sign measurements throughout the stay

## 0. Setup

In [1]:
import pandas as pd
import os

# Path to the data folder
data_path = r"mimic-iv-ed-2.2\ed"

print("Setup complete. Data path:", data_path)

Setup complete. Data path: mimic-iv-ed-2.2\ed


---
## 1. Table-by-Table Exploration

For each table we will inspect:
- The first rows (`.head()`)
- Total number of rows
- Number of unique patients (`subject_id`)
- Number of unique stays (`stay_id`)
- Top 15 most frequent values of the most relevant categorical column

### 1.1 Diagnosis

In [2]:
df_diagnosis = pd.read_csv(os.path.join(data_path, "diagnosis.csv.gz"))

print(f"Diagnosis table has {df_diagnosis.shape[0]:,} rows and {df_diagnosis.shape[1]} columns.\n")
display(df_diagnosis.head())

# Count of unique IDs
patients_diagnosis = df_diagnosis['subject_id'].nunique()
visits_diagnosis = df_diagnosis['stay_id'].nunique()
print(f"Unique patients: {patients_diagnosis:,}")
print(f"Unique emergency visits (Stays): {visits_diagnosis:,}\n")

# Top 15 most frequent diagnoses
print("Top 15 most common diagnoses (ICD Title):")
display(df_diagnosis['icd_title'].value_counts().head(15))

Diagnosis table has 899,050 rows and 6 columns.



,subject_id,stay_id,seq_num,icd_code,icd_version,icd_title
0,10000032,32952584,1,4589,9,HYPOTENSION NOS
1,10000032,32952584,2,07070,9,UNSPECIFIED VIRAL HEPATITIS C WITHOUT HEPATIC ...
2,10000032,32952584,3,V08,9,ASYMPTOMATIC HIV INFECTION
3,10000032,33258284,1,5728,9,"OTH SEQUELA, CHR LIV DIS"
4,10000032,33258284,2,78959,9,OTHER ASCITES


Unique patients: 205,129
Unique emergency visits (Stays): 423,989

Top 15 most common diagnoses (ICD Title):


icd_title
HYPERTENSION NOS                                      26816
Essential (primary) hypertension                      21264
Chest pain, unspecified                               13016
CHEST PAIN NOS                                        12398
DIABETES UNCOMPL ADULT                                12026
Unspecified abdominal pain                            10504
Type 2 diabetes mellitus without complications         8801
ABDOMINAL PAIN OTHER SPECIED                           8704
Fall on same level, unspecified, initial encounter     7817
UNSPECIFIED FALL                                       7081
Dyspnea, unspecified                                   7080
URIN TRACT INFECTION NOS                               6877
Alcohol abuse with intoxication, unspecified           6871
Headache                                               6841
HYPERCHOLESTEROLEMIA                                   6662
Name: count, dtype: int64

From the extracted data, the table contains a large number of entries (899,050 rows) belonging to a smaller number of unique patients (205,129) and unique hospital stays (423,989). This indicates that a single patient typically receives **multiple diagnoses during the same visit**.

### 1.2 Edstays

In [3]:
df_edstays = pd.read_csv(os.path.join(data_path, "edstays.csv.gz"))

print(f"Edstays table has {df_edstays.shape[0]:,} rows and {df_edstays.shape[1]} columns.\n")
display(df_edstays.head())

# Count of unique IDs
patients_edstays = df_edstays['subject_id'].nunique()
visits_edstays = df_edstays['stay_id'].nunique()
print(f"Unique patients: {patients_edstays:,}")
print(f"Unique emergency visits (Stays): {visits_edstays:,}\n")

# Top 15 arrival transports
print("Top 15 arrival transports:")
display(df_edstays['arrival_transport'].value_counts().head(15))

Edstays table has 425,087 rows and 9 columns.



,subject_id,hadm_id,stay_id,intime,outtime,gender,race,arrival_transport,disposition
0,10000032,22595853.0,33258284,2180-05-06 19:17:00,2180-05-06 23:30:00,F,WHITE,AMBULANCE,ADMITTED
1,10000032,22841357.0,38112554,2180-06-26 15:54:00,2180-06-26 21:31:00,F,WHITE,AMBULANCE,ADMITTED
2,10000032,25742920.0,35968195,2180-08-05 20:58:00,2180-08-06 01:44:00,F,WHITE,AMBULANCE,ADMITTED
3,10000032,29079034.0,32952584,2180-07-22 16:24:00,2180-07-23 05:54:00,F,WHITE,AMBULANCE,HOME
4,10000032,29079034.0,39399961,2180-07-23 05:54:00,2180-07-23 14:00:00,F,WHITE,AMBULANCE,ADMITTED


Unique patients: 205,504
Unique emergency visits (Stays): 425,087

Top 15 arrival transports:


arrival_transport
WALK IN       251849
AMBULANCE     155752
UNKNOWN        15352
OTHER           1266
HELICOPTER       868
Name: count, dtype: int64

### 1.3 Triage

In [4]:
df_triage = pd.read_csv(os.path.join(data_path, "triage.csv.gz"))

print(f"Triage table has {df_triage.shape[0]:,} rows and {df_triage.shape[1]} columns.\n")
display(df_triage.head())

# Count of unique IDs
patients_triage = df_triage['subject_id'].nunique()
visits_triage = df_triage['stay_id'].nunique()
print(f"Unique patients: {patients_triage:,}")
print(f"Unique emergency visits (Stays): {visits_triage:,}\n")

# Top 15 chief complaints
print("Top 15 chief complaints:")
display(df_triage['chiefcomplaint'].value_counts().head(15))

Triage table has 425,087 rows and 11 columns.



,subject_id,stay_id,temperature,heartrate,resprate,o2sat,sbp,dbp,pain,acuity,chiefcomplaint
0,10000032,32952584,97.8,87.0,14.0,97.0,71.0,43.0,7,2.0,Hypotension
1,10000032,33258284,98.4,70.0,16.0,97.0,106.0,63.0,0,3.0,"Abd pain, Abdominal distention"
2,10000032,35968195,99.4,105.0,18.0,96.0,106.0,57.0,10,3.0,"n/v/d, Abd pain"
3,10000032,38112554,98.9,88.0,18.0,97.0,116.0,88.0,10,3.0,Abdominal distention
4,10000032,39399961,98.7,77.0,16.0,98.0,96.0,50.0,13,2.0,"Abdominal distention, Abd pain, LETHAGIC"


Unique patients: 205,504
Unique emergency visits (Stays): 425,087

Top 15 chief complaints:


chiefcomplaint
Chest pain               11301
Abd pain                 10771
Dyspnea                   6003
s/p Fall                  4965
ABD PAIN                  4957
SI                        4926
ETOH                      4679
Wound eval                4509
Headache                  3828
Back pain                 3166
MVC                       3134
Altered mental status     2688
Lower back pain           2630
S/P FALL                  2579
CHEST PAIN                2577
Name: count, dtype: int64

### 1.4 Medrecon (Medication Reconciliation)

In [5]:
df_medrecon = pd.read_csv(os.path.join(data_path, "medrecon.csv.gz"))

print(f"Medrecon table has {df_medrecon.shape[0]:,} rows and {df_medrecon.shape[1]} columns.\n")
display(df_medrecon.head())

# Count of unique IDs
patients_medrecon = df_medrecon['subject_id'].nunique()
visits_medrecon = df_medrecon['stay_id'].nunique()
print(f"Unique patients: {patients_medrecon:,}")
print(f"Unique emergency visits (Stays): {visits_medrecon:,}\n")

# Top 15 prior medications
print("Top 15 most common prior medications:")
display(df_medrecon['name'].value_counts().head(15))

Medrecon table has 2,987,342 rows and 9 columns.



,subject_id,stay_id,charttime,name,gsn,ndc,etc_rn,etccode,etcdescription
0,10000032,32952584,2180-07-22 17:26:00,albuterol sulfate,28090,21695042308,1,5970.0,Asthma/COPD Therapy - Beta 2-Adrenergic Agents...
1,10000032,32952584,2180-07-22 17:26:00,calcium carbonate,1340,10135021101,1,733.0,Minerals and Electrolytes - Calcium Replacement
2,10000032,32952584,2180-07-22 17:26:00,cholecalciferol (vitamin D3),65241,37205024678,1,670.0,Vitamins - D Derivatives
3,10000032,32952584,2180-07-22 17:26:00,emtricitabine-tenofovir [Truvada],57883,35356007003,1,5849.0,Antiretroviral - Nucleoside and Nucleotide Ana...
4,10000032,32952584,2180-07-22 17:26:00,fluticasone [Flovent HFA],21251,49999061401,1,371.0,Asthma Therapy - Inhaled Corticosteroids (Gluc...


Unique patients: 141,464
Unique emergency visits (Stays): 307,196

Top 15 most common prior medications:


name
aspirin                         105646
omeprazole                       55518
lisinopril                       48669
lorazepam                        41402
clonazepam                       40203
gabapentin                       36589
atorvastatin                     35618
cholecalciferol (vitamin D3)     33186
amlodipine                       32929
metoprolol succinate             31941
acetaminophen                    29336
levothyroxine                    28411
metformin                        27943
simvastatin                      27249
albuterol sulfate                27245
Name: count, dtype: int64

### 1.5 Pyxis (Dispensed Medications inside ED)

In [6]:
df_pyxis = pd.read_csv(os.path.join(data_path, "pyxis.csv.gz"))

print(f"Pyxis table has {df_pyxis.shape[0]:,} rows and {df_pyxis.shape[1]} columns.\n")
display(df_pyxis.head())

# Count of unique IDs
patients_pyxis = df_pyxis['subject_id'].nunique()
visits_pyxis = df_pyxis['stay_id'].nunique()
print(f"Unique patients: {patients_pyxis:,}")
print(f"Unique emergency visits (Stays): {visits_pyxis:,}\n")

# Top 15 dispensed medications
print("Top 15 most common dispensed medications:")
display(df_pyxis['name'].value_counts().head(15))

Pyxis table has 1,586,053 rows and 7 columns.



,subject_id,stay_id,charttime,med_rn,name,gsn_rn,gsn
0,10000032,32952584,2180-07-22 17:59:00,1,Albuterol Inhaler,1,5037.0
1,10000032,32952584,2180-07-22 17:59:00,1,Albuterol Inhaler,2,28090.0
2,10000032,35968195,2180-08-05 22:29:00,1,Morphine,1,4080.0
3,10000032,35968195,2180-08-05 22:55:00,2,Donnatol (Elixir),1,4773.0
4,10000032,35968195,2180-08-05 22:55:00,3,Aluminum-Magnesium Hydrox.-Simet,1,2701.0


Unique patients: 156,768
Unique emergency visits (Stays): 303,361

Top 15 most common dispensed medications:


name
Ondansetron                          84890
Acetaminophen                        45740
Lorazepam                            43798
Ondansetron 4mg/2mL 2mL VIAL         41162
Morphine Sulfate                     39854
HYDROmorphone (Dilaudid)             37616
Morphine                             29855
Vancomycin                           27226
Acetaminophen 500mg TAB              26746
Morphine Sulfat 4mg/1mL 1mL VIAL     26157
LORazepam 2mg/1mL 1mL SYR            24762
Tetanus-DiphTox-Acellular Pertuss    22283
Ibuprofen                            18881
Ketorolac                            16908
Diazepam                             16207
Name: count, dtype: int64

### 1.6 Vitalsign

In [7]:
df_vitalsign = pd.read_csv(os.path.join(data_path, "vitalsign.csv.gz"))

print(f"Vitalsign table has {df_vitalsign.shape[0]:,} rows and {df_vitalsign.shape[1]} columns.\n")
display(df_vitalsign.head())

# Count of unique IDs
patients_vitalsign = df_vitalsign['subject_id'].nunique()
visits_vitalsign = df_vitalsign['stay_id'].nunique()
print(f"Unique patients: {patients_vitalsign:,}")
print(f"Unique emergency visits (Stays): {visits_vitalsign:,}\n")

# Note: vitalsign contains numerical data only (heartrate, temperature, etc.)
print("Columns available:")
print(df_vitalsign.columns.tolist())

Vitalsign table has 1,564,610 rows and 11 columns.



,subject_id,stay_id,charttime,temperature,heartrate,resprate,o2sat,sbp,dbp,rhythm,pain
0,10000032,32952584,2180-07-22 16:36:00,NaN,83.0,24.0,97.0,90.0,51.0,NaN,0
1,10000032,32952584,2180-07-22 16:43:00,NaN,85.0,22.0,98.0,76.0,39.0,NaN,0
2,10000032,32952584,2180-07-22 16:45:00,NaN,84.0,22.0,97.0,75.0,39.0,NaN,0
3,10000032,32952584,2180-07-22 17:56:00,NaN,84.0,20.0,99.0,86.0,51.0,NaN,NaN
4,10000032,32952584,2180-07-22 18:37:00,98.4,86.0,20.0,98.0,65.0,37.0,NaN,NaN


Unique patients: 198,131
Unique emergency visits (Stays): 408,146

Columns available:
['subject_id', 'stay_id', 'charttime', 'temperature', 'heartrate', 'resprate', 'o2sat', 'sbp', 'dbp', 'rhythm', 'pain']


---
## 2. Comparative Summary

A consolidated view of all tables: total rows, unique patients and unique visits.

In [8]:
comparative_data = {
    "Table Name": ["Diagnosis", "Edstays", "Triage", "Medrecon", "Pyxis", "Vitalsign"],
    "Total Rows": [
        f"{df_diagnosis.shape[0]:,}", f"{df_edstays.shape[0]:,}", f"{df_triage.shape[0]:,}",
        f"{df_medrecon.shape[0]:,}", f"{df_pyxis.shape[0]:,}", f"{df_vitalsign.shape[0]:,}"
    ],
    "Unique Patients": [
        f"{patients_diagnosis:,}", f"{patients_edstays:,}", f"{patients_triage:,}",
        f"{patients_medrecon:,}", f"{patients_pyxis:,}", f"{patients_vitalsign:,}"
    ],
    "Unique Visits (stay_id)": [
        f"{visits_diagnosis:,}", f"{visits_edstays:,}", f"{visits_triage:,}",
        f"{visits_medrecon:,}", f"{visits_pyxis:,}", f"{visits_vitalsign:,}"
    ]
}

df_comparative = pd.DataFrame(comparative_data)
display(df_comparative)

,Table Name,Total Rows,Unique Patients,Unique Visits (stay_id)
0,Diagnosis,"899,050","205,129","423,989"
1,Edstays,"425,087","205,504","425,087"
2,Triage,"425,087","205,504","425,087"
3,Medrecon,"2,987,342","141,464","307,196"
4,Pyxis,"1,586,053","156,768","303,361"
5,Vitalsign,"1,564,610","198,131","408,146"


---
## 3. Conclusions
### GEMINI
Basándonos en nuestra tabla comparativa de volumen, podemos inferir varias conclusiones fundamentales sobre la estructura de nuestra base de datos que dictarán cómo trataremos las variables antes de modelar:

1. **`Edstays` y `Triage` son la base perfecta (Core Tables):**
   Ambas tablas comparten exactamente la misma cantidad de filas (425,087), visitas únicas (425,087) y pacientes únicos (205,504). Esto nos garantiza matemáticamente que hay una relación de **1 a 1**; es decir, todo paciente que es admitido en urgencias pasa por un único triaje inicial obligatorio. *(`Edstays` será nuestra tabla "ancla" para calcular la variable objetivo de readmisión).*

2. **La Tasa Base de Pacientes Recurrentes:**
   Los `205,504` pacientes únicos generan un total de `425,087` estancias. Esto implica un promedio general de unas **2 visitas a urgencias por paciente**, lo cual nos adelanta que el fenómeno de la **readmisión** está fuertemente presente en nuestra base de datos.

3. **Múltiples diagnósticos por estancia (`Diagnosis`):**
   Aunque hay casi 424,000 estancias diagnosticadas, los registros se elevan a casi `900,000` filas. Esto confirma que el médico promedio suele asignar **más de un código ICD (diagnóstico)** a un paciente durante una misma visita a emergencias (por ejemplo, el motivo principal más una patología crónica).

4. **El historial farmacológico masivo (`Medrecon`):**
   Esta tabla posee un abrumador total de casi `3 millones` de filas para tan solo `307,196` visitas. Esto indica que los pacientes suelen llegar con listas de prescripciones previas bastante extensas (una media aproximada de 10 medicamentos crónicos por persona). Sin embargo, notablemente, **solo cubre a ~141,000 pacientes**, revelando que miles de personas acuden a urgencias sin historial previo de medicación.

5. **Dispensación clínica y Monitorización (`Pyxis` y `Vitalsign`):**
   Ambas tablas operan bajo un esquema de "múltiples tomas". Un paciente (un `stay_id`) recibe varias rondas de constantes vitales (`Vitalsign` reporta 1.5M de filas) y múltiples dispensaciones de medicamentos (`Pyxis`). Tendremos que agrupar (agregar usando promedios o *min/max*) estas filas por estancia antes de unirlas a nuestra tabla principal, porque de lo contrario duplicaríamos un mismo paciente cientos de veces en nuestro dataset de Machine Learning.

## CLAUDE

Based on the comparative table, we can draw the following key conclusions:

1. **`Edstays` and `Triage` are the core tables (1-to-1 relationship):**  
   Both share exactly the same number of rows (425,087), unique patients (205,504) and unique stays (425,087). Every admitted patient undergoes exactly one triage. `Edstays` will serve as our **anchor table** when building the readmission target variable.

2. **High patient recurrence rate:**  
   205,504 unique patients generate 425,087 total stays, meaning an average of ~2 ED visits per patient. This confirms that **readmission is a frequent and relevant phenomenon** in this dataset.

3. **Multiple diagnoses per stay (`Diagnosis`):**  
   899,050 rows for ~424,000 unique stays means physicians assign on average **more than 2 ICD codes per visit** (primary condition + chronic comorbidities).

4. **Large medication history (`Medrecon`):**  
   Nearly 3 million rows for ~307,000 stays indicates patients arrive with extensive prior prescription lists (~10 chronic medications on average). Notably, this table only covers ~141,000 patients, meaning many patients have **no prior medication history** recorded.

5. **Repeated measurements require aggregation (`Pyxis` and `Vitalsign`):**  
   Both tables contain multiple rows per stay (many dispensations and vital sign readings per visit). Before merging them into a master dataset for modeling, we will need to **aggregate these rows per `stay_id`** (e.g., using mean, min, max) to avoid duplicating patients.